File: php/src/MlDsa/DsaHash.php
<?php

declare(strict_types=1);

namespace PQC\MlDsa;

use PQC\MlKem\HashFuncs;

/**
 * Hash functions for ML-DSA.
 */
final class DsaHash
{
    /**
     * H (SHAKE-256 with variable output).
     */
    public static function H(string $input, int $outLen): string
    {
        return HashFuncs::shake256($input, $outLen);
    }

    /**
     * SHAKE-128 XOF.
     */
    public static function shake128(string $input, int $outLen): string
    {
        return HashFuncs::shake128($input, $outLen);
    }

    /**
     * SHAKE-256 XOF.
     */
    public static function shake256(string $input, int $outLen): string
    {
        return HashFuncs::shake256($input, $outLen);
    }

    /**
     * ExpandA: sample matrix A from rho using SHAKE-128.
     * FIPS 204 Algorithm 26.
     */
    public static function expandA(string $rho, int $k, int $l): array
    {
        $A = [];
        for ($r = 0; $r < $k; $r++) {
            for ($s = 0; $s < $l; $s++) {
                $A[$r][$s] = self::rejNttPoly($rho . chr($s) . chr($r));
            }
        }
        return $A;
    }

    /**
     * RejNTTPoly: rejection sampling to produce NTT-domain polynomial.
     * FIPS 204 Algorithm 24.
     */
    public static function rejNttPoly(string $seed): array
    {
        $stream = self::shake128($seed, 256 * 4);
        $coeffs = [];
        $pos = 0;

        while (count($coeffs) < 256) {
            if ($pos + 3 > strlen($stream)) {
                $stream .= self::shake128($seed . pack('V', $pos), 256 * 4);
            }
            $b0 = ord($stream[$pos++]);
            $b1 = ord($stream[$pos++]);
            $b2 = ord($stream[$pos++]);

            $val = $b0 | ($b1 << 8) | ($b2 << 16);
            $val &= 0x7FFFFF; // 23 bits

            if ($val < DsaField::Q) {
                $coeffs[] = $val;
            }
        }

        return array_slice($coeffs, 0, 256);
    }

    /**
     * ExpandS: sample secret vectors s1, s2 from rhoPrime.
     * FIPS 204 Algorithm 27.
     */
    public static function expandS(string $rhoPrime, int $eta, int $k, int $l): array
    {
        $s1 = [];
        for ($r = 0; $r < $l; $r++) {
            $s1[$r] = self::rejBoundedPoly($rhoPrime . pack('v', $r), $eta);
        }

        $s2 = [];
        for ($r = 0; $r < $k; $r++) {
            $s2[$r] = self::rejBoundedPoly($rhoPrime . pack('v', $l + $r), $eta);
        }

        return [$s1, $s2];
    }

    /**
     * RejBoundedPoly: sample polynomial with coefficients in [-eta, eta].
     * FIPS 204 Algorithm 25.
     */
    public static function rejBoundedPoly(string $seed, int $eta): array
    {
        $stream = self::shake256($seed, 256 * 2);
        $coeffs = [];
        $pos = 0;

        while (count($coeffs) < 256) {
            if ($pos >= strlen($stream)) {
                $stream .= self::shake256($seed . pack('V', $pos), 256 * 2);
            }
            $byte = ord($stream[$pos++]);

            $z0 = $byte & 0x0F;
            $z1 = $byte >> 4;

            if ($eta === 2) {
                if ($z0 < 15) {
                    $val = $z0 - (int)(($z0 * 205) >> 10) * 5;
                    $coeffs[] = DsaField::mod($eta - $val);
                }
                if (count($coeffs) < 256 && $z1 < 15) {
                    $val = $z1 - (int)(($z1 * 205) >> 10) * 5;
                    $coeffs[] = DsaField::mod($eta - $val);
                }
            } else { // eta = 4
                if ($z0 < 9) {
                    $coeffs[] = DsaField::mod($eta - $z0);
                }
                if (count($coeffs) < 256 && $z1 < 9) {
                    $coeffs[] = DsaField::mod($eta - $z1);
                }
            }
        }

        return array_slice($coeffs, 0, 256);
    }

    /**
     * ExpandMask: sample mask vector y from rhoPrime and kappa.
     * FIPS 204 Algorithm 28.
     */
    public static function expandMask(string $rhoPrime, int $kappa, int $l, int $gamma1): array
    {
        $y = [];
        $gamma1Bits = ($gamma1 === (1 << 17)) ? 18 : 20;

        for ($r = 0; $r < $l; $r++) {
            $stream = self::shake256(
                $rhoPrime . pack('v', $kappa + $r),
                32 * $gamma1Bits
            );
            $y[$r] = self::bitUnpackGamma1($stream, $gamma1, $gamma1Bits);
        }

        return $y;
    }

    /**
     * Unpack gamma1-bounded coefficients from byte stream.
     */
    private static function bitUnpackGamma1(string $stream, int $gamma1, int $bits): array
    {
        $coeffs = [];
        if ($bits === 18) {
            // 9 bytes -> 4 coefficients
            for ($i = 0; $i < 64; $i++) {
                $offset = $i * 9;
                $b = [];
                for ($j = 0; $j < 9; $j++) {
                    $b[$j] = ord($stream[$offset + $j]);
                }

                $val0 = $b[0] | ($b[1] << 8) | (($b[2] & 0x03) << 16);
                $val1 = (($b[2] >> 2) | ($b[3] << 6) | (($b[4] & 0x0F) << 14));
                $val2 = (($b[4] >> 4) | ($b[5] << 4) | (($b[6] & 0x3F) << 12));
                $val3 = (($b[6] >> 6) | ($b[7] << 2) | ($b[8] << 10));

                $val0 &= 0x3FFFF;
                $val1 &= 0x3FFFF;
                $val2 &= 0x3FFFF;
                $val3 &= 0x3FFFF;

                $coeffs[] = DsaField::mod($gamma1 - $val0);
                $coeffs[] = DsaField::mod($gamma1 - $val1);
                $coeffs[] = DsaField::mod($gamma1 - $val2);
                $coeffs[] = DsaField::mod($gamma1 - $val3);
            }
        } else { // 20 bits
            // 5 bytes -> 2 coefficients
            for ($i = 0; $i < 128; $i++) {
                $offset = $i * 5;
                $b0 = ord($stream[$offset]);
                $b1 = ord($stream[$offset + 1]);
                $b2 = ord($stream[$offset + 2]);
                $b3 = ord($stream[$offset + 3]);
                $b4 = ord($stream[$offset + 4]);

                $val0 = $b0 | ($b1 << 8) | (($b2 & 0x0F) << 16);
                $val1 = ($b2 >> 4) | ($b3 << 4) | ($b4 << 12);

                $val0 &= 0xFFFFF;
                $val1 &= 0xFFFFF;

                $coeffs[] = DsaField::mod($gamma1 - $val0);
                $coeffs[] = DsaField::mod($gamma1 - $val1);
            }
        }

        return array_slice($coeffs, 0, 256);
    }

    /**
     * SampleInBall: generate challenge polynomial c with tau +/-1 coefficients.
     * FIPS 204 Algorithm 29.
     */
    public static function sampleInBall(string $seed, int $tau): array
    {
        $stream = self::shake256($seed, 8 + 256);
        $c = array_fill(0, 256, 0);

        // Extract sign bits from first 8 bytes
        $signs = 0;
        for ($i = 7; $i >= 0; $i--) {
            $signs = ($signs << 8) | ord($stream[$i]);
        }

        $pos = 8;
        for ($i = 256 - $tau; $i < 256; $i++) {
            // Get j from stream
            while (true) {
                if ($pos >= strlen($stream)) {
                    $stream .= self::shake256($seed . pack('V', $pos), 256);
                }
                $j = ord($stream[$pos++]);
                if ($j <= $i) break;
            }

            $c[$i] = $c[$j];
            $c[$j] = (($signs & 1) === 0) ? 1 : DsaField::mod(-1);
            $signs >>= 1;
        }

        return $c;
    }
}

File: dotnet/src/PqcStandards/MlDsa/DsaHash.cs
using PqcStandards.Common;

namespace PqcStandards.MlDsa;

/// <summary>Hash/sampling functions for ML-DSA (FIPS 204).</summary>
public static class DsaHash
{
    /// <summary>ExpandA: generate k x l matrix of NTT-domain polynomials from rho.</summary>
    public static int[][] ExpandA(byte[] rho, int k, int l)
    {
        int[][] A = new int[k * l][];
        for (int i = 0; i < k; i++)
            for (int j = 0; j < l; j++)
                A[i * l + j] = RejNttPoly(rho, (byte)j, (byte)i);
        return A;
    }

    private static int[] RejNttPoly(byte[] rho, byte j, byte i)
    {
        byte[] input = new byte[rho.Length + 2];
        Buffer.BlockCopy(rho, 0, input, 0, rho.Length);
        input[rho.Length] = j;
        input[rho.Length + 1] = i;

        byte[] buf = Keccak.Shake128(input, 3 * 256 * 2);
        int[] a = new int[256];
        int ctr = 0, pos = 0;
        while (ctr < 256)
        {
            int val = (buf[pos] | (buf[pos + 1] << 8) | (buf[pos + 2] << 16)) & 0x7FFFFF;
            pos += 3;
            if (val < DsaField.Q)
                a[ctr++] = val;
        }
        return a;
    }

    /// <summary>ExpandS: generate secret vectors s1 (l polys) and s2 (k polys) from rhoPrime.</summary>
    public static (int[][] s1, int[][] s2) ExpandS(byte[] rhoPrime, int k, int l, int eta)
    {
        int[][] s1 = new int[l][];
        int[][] s2 = new int[k][];
        ushort nonce = 0;
        for (int i = 0; i < l; i++)
            s1[i] = RejBoundedPoly(rhoPrime, nonce++, eta);
        for (int i = 0; i < k; i++)
            s2[i] = RejBoundedPoly(rhoPrime, nonce++, eta);
        return (s1, s2);
    }

    private static int[] RejBoundedPoly(byte[] rhoPrime, ushort nonce, int eta)
    {
        byte[] input = new byte[rhoPrime.Length + 2];
        Buffer.BlockCopy(rhoPrime, 0, input, 0, rhoPrime.Length);
        input[rhoPrime.Length] = (byte)(nonce & 0xFF);
        input[rhoPrime.Length + 1] = (byte)(nonce >> 8);

        int outLen = eta <= 2 ? 128 * 3 : 128 * 3;
        byte[] buf = Keccak.Shake256(input, outLen);

        int[] p = new int[256];
        int ctr = 0, pos = 0;
        while (ctr < 256 && pos < buf.Length)
        {
            if (eta == 2)
            {
                int b = buf[pos++];
                int z0 = b & 0x0F;
                int z1 = b >> 4;
                if (z0 < 15) { p[ctr++] = DsaField.ModQ(CoeffFromHalfByte(z0, eta)); }
                if (ctr < 256 && z1 < 15) { p[ctr++] = DsaField.ModQ(CoeffFromHalfByte(z1, eta)); }
            }
            else // eta == 4
            {
                int b = buf[pos++];
                int z0 = b & 0x0F;
                int z1 = b >> 4;
                if (z0 < 9) { p[ctr++] = DsaField.ModQ(4 - z0); }
                if (ctr < 256 && z1 < 9) { p[ctr++] = DsaField.ModQ(4 - z1); }
            }
        }
        return p;
    }

    private static int CoeffFromHalfByte(int z, int eta)
    {
        if (eta == 2)
        {
            int r = z % 5;
            return 2 - r;
        }
        return 4 - z;
    }

    /// <summary>ExpandMask: generate masking vector y from rhoPrime and kappa.</summary>
    public static int[][] ExpandMask(byte[] rhoPrime, int kappa, int l, int gamma1)
    {
        int[][] y = new int[l][];
        int bitLen = gamma1 == (1 << 17) ? 18 : 20;
        int polyBytes = 32 * bitLen;
        for (int i = 0; i < l; i++)
        {
            ushort nonce = (ushort)(kappa + i);
            byte[] input = new byte[rhoPrime.Length + 2];
            Buffer.BlockCopy(rhoPrime, 0, input, 0, rhoPrime.Length);
            input[rhoPrime.Length] = (byte)(nonce & 0xFF);
            input[rhoPrime.Length + 1] = (byte)(nonce >> 8);
            byte[] buf = Keccak.Shake256(input, polyBytes);
            y[i] = BitUnpackGamma(buf, gamma1, bitLen);
        }
        return y;
    }

    private static int[] BitUnpackGamma(byte[] buf, int gamma1, int bitLen)
    {
        int[] poly = new int[256];
        for (int i = 0; i < 256; i++)
        {
            int bitPos = i * bitLen;
            long val = 0;
            for (int b = 0; b < bitLen; b++)
            {
                int idx = (bitPos + b) / 8;
                int bit = (bitPos + b) % 8;
                if (idx < buf.Length)
                    val |= (long)((buf[idx] >> bit) & 1) << b;
            }
            poly[i] = DsaField.ModQ(gamma1 - (int)val);
        }
        return poly;
    }

    /// <summary>SampleInBall: deterministic sampling of challenge polynomial c.</summary>
    public static int[] SampleInBall(byte[] seed, int tau)
    {
        byte[] buf = Keccak.Shake256(seed, 8 + 256);
        int[] c = new int[256];

        // Extract 8 bytes for sign bits
        ulong signs = 0;
        for (int i = 0; i < 8; i++)
            signs |= (ulong)buf[i] << (8 * i);

        int k = 8;
        for (int i = 256 - tau; i < 256; i++)
        {
            int j;
            do
            {
                j = buf[k++];
                if (k >= buf.Length)
                {
                    buf = Keccak.Shake256(buf, 8 + 256);
                    k = 0;
                }
            } while (j > i);

            c[i] = c[j];
            c[j] = ((signs & 1) == 1) ? DsaField.Q - 1 : 1;
            signs >>= 1;
        }
        return c;
    }
}
